File: Java/src/stocks/Stock.java
package stocks;

import java.util.ArrayList;
import java.util.List;

public class Stock {

	private String name;
	private List<Candle> candles = new ArrayList<>();
	private List<Double> ma = new ArrayList<>();
	private List<Double> ema = new ArrayList<>();
	
	public Stock(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}

	public List<Candle> getCandles() {
		return candles;
	}

	public List<Double> getMA() {
		return ma;
	}

	public void clearMA() {
		ma.clear();
	}

	public List<Double> getEMA() {
		return ema;
	}

	public void clearEMA() {
		ema.clear();;
	}

	public void addCandle(Candle candle) {
		candles.add(candle);		
	}
	
	public void calculateMA(int n) {
		ma.clear();
		int num, j; double v, sum;
		for (int i = 0; i < candles.size(); i++) {
			sum = 0;
			j = i;
			for(num = n + 1; num > 0; num--) {
				sum += candles.get(j).getClose();
				if (j == 0) {
					num--; 
					break;
				}
				else j--;
			}
			v = sum / (n + 1 - num);
			ma.add(v);
		}
	}
	
	public void calculateEMA(int n) {
		ema.clear();
		List<Candle> c = new ArrayList<>(); 
		double v; int num, j; 
		for (int i = 0; i < candles.size(); i++) {
			j = i;
			for (num = n + 1; num > 0; num--) {
				c.add(candles.get(j));
				if (j == 0) break;
				else j--;
			}
			v = ema(c.size(), c);
			ema.add(v);
			c.clear();
		}
	}
	
	public double ema(int n, List<Candle> c) {
		double s = 0;
		if (n == 1) return c.get(0).getClose();		
		s += c.get(n-1).getClose() * (2. / (n + 1)) + ema((n - 1), c) * (1 - (2. / (n + 1)));
		return s;
	}

}

File: C++/stock.cpp
#include "stock.h"

ostream & operator<<(ostream & os, const Stock & s) {
	vector<double>::iterator j; vector<double>::iterator k;
	os << s.name << endl;
	os << left<< setw(13) <<"Timestamp" << left << setw(14) << "Open" << left << setw(14) << "Close";
	os << left << setw(14) << "High" << left << setw(14) << "Low"; 	
	if (s.ma) {
		j = s.ma->getValues()->begin();
		os << "MA(" << s.ma->getN() << ")         ";
	}
	if (s.ema) {
		k = s.ema->getValues()->begin();
		os << "EMA(" << s.ema->getN() << ")";
	}
	os << endl;
	for (Candle* c : s.candles) {
		os << *c;
		if (s.ma) { os << left << setw(14) << setprecision(10) << *j; j++; }
		if (s.ema) { os << left << setw(14) << setprecision(10) << *k; k++; }
		os << endl;
	}
	return os;
}

void Stock::copy(const Stock & s) {
	this->name = s.name;
	for_each(s.candles.begin(), s.candles.end(), [this, s](Candle *c) {
		Candle* nc = new Candle(*c);
		this->addCandle(nc);
	});
	this->ma = new MA(*(s.ma));
	this->ema = new EMA(*(s.ema));
}

void Stock::move(Stock & s) {
	this->name = s.name;
	for_each(s.candles.begin(), s.candles.end(), [this, s](Candle* c){
		this->candles.push_back(c);
		c = nullptr;
	});
	s.candles.clear();
	this->ma = s.ma; this->ema = s.ema;
	s.ma = nullptr; s.ema = nullptr;
}

void Stock::remove() {
	for_each(candles.begin(), candles.end(), [](Candle *c) {
		delete c;
	});
	candles.clear();
	delete ma; 
	delete ema;
}

Stock & Stock::operator=(const Stock & s) {
	if (this != &s) {
		remove();
		copy(s);
	}
	return *this;
}

Stock & Stock::operator=(Stock && s) {
	if (this != &s) {
		remove();
		move(s);
	}
	return *this;
}

void Stock::addCandle(Candle * c) {
	candles.push_back(c);
}

int Stock::numOfCandles() const {
	return candles.size();
}

void Stock::calculateMA(int n) {
	ma = new MA(n);
	ma->calculate(getCandles());
}

void Stock::calculateEMA(int n) {
	ema = new EMA(n);
	ema->calculate(getCandles());
}
